src/ucs/datastruct/arbiter.c

```diff
@@ -170,207 +150,178 @@ void ucs_arbiter_group_purge(ucs_arbiter_t *arbiter,
+                                           ucs_arbiter_elem_t *head)
+{
+    if (!ucs_arbiter_group_head_is_scheduled(head)) {
+        ucs_list_add_tail(&arbiter->list, &head->list);
     }
 }
 
 void ucs_arbiter_group_schedule_nonempty(ucs_arbiter_t *arbiter,
                                          ucs_arbiter_group_t *group)
 {
     ucs_arbiter_elem_t *tail = group->tail;
-    ucs_arbiter_elem_t *current, *head;
-
-    UCS_ARBITER_GUARD_CHECK(arbiter);
+    ucs_arbiter_elem_t *head;
 
     ucs_assert(tail != NULL);
     head = tail->next;
 
     if (head == NULL) {
-        /* it means that 1 element group is
-         * scheduled during dispatch.
+        /* It means that 1 element group is scheduled during dispatch.
          * Restore next pointer.
          */
-        head = tail->next = tail;
-    }
-
-    if (ucs_arbiter_group_is_scheduled(head)) {
-        return; /* Already scheduled */
+        head = tail;
     }
 
-    current = arbiter->current;
-    if (current == NULL) {
-        ucs_list_head_init(&head->list);
-        arbiter->current = head;
-    } else {
-        ucs_list_insert_before(&current->list, &head->list);
-    }
+    ucs_arbiter_schedule_head_if_not_scheduled(arbiter, head);
 }
 
 void ucs_arbiter_dispatch_nonempty(ucs_arbiter_t *arbiter, unsigned per_group,
                                    ucs_arbiter_callback_t cb, void *cb_arg)
 {
-    ucs_arbiter_elem_t *group_head, *last_elem, *elem, *next_elem;
-    ucs_list_link_t *elem_list_next;
-    ucs_arbiter_elem_t *next_group, *prev_group;
-    ucs_arbiter_group_t *group;
+    ucs_arbiter_elem_t *group_head, *group_tail, *next_elem;
     ucs_arbiter_cb_result_t result;
     unsigned group_dispatch_count;
-    int is_single_group;
-    UCS_LIST_HEAD(resched_groups);
+    ucs_arbiter_group_t *group;
+    UCS_LIST_HEAD(resched_list);
+    int sched_group;
 
-    next_group = arbiter->current;
-    ucs_assert(next_group != NULL);
+    ucs_assert(!ucs_list_is_empty(&arbiter->list));
 
-    do {
-        group_head    = next_group;
+    for (;;) {
+        group_head = ucs_list_extract_head(&arbiter->list, ucs_arbiter_elem_t,
+                                           list);
         ucs_assert(group_head != NULL);
-        prev_group    = ucs_list_prev(&group_head->list, ucs_arbiter_elem_t, list);
-        next_group    = ucs_list_next(&group_head->list, ucs_arbiter_elem_t, list);
-        ucs_assert(prev_group != NULL);
-        ucs_assert(next_group != NULL);
-        ucs_assert(prev_group->list.next == &group_head->list);
-        ucs_assert(next_group->list.prev == &group_head->list);
+
+        /* TODO we should disallow group scheduling while the group is being
+         * dispatched: it will allow removing the next statement and reset the
+         * head only in case DESCHED is returned, and also remove the head==NULL
+         * check from ucs_arbiter_group_schedule_nonempty(). Seems only DC is
```

Comment:
when it is scheduled during dispatch with DC? Is it when moving from waitq to txq with dci policy?